File: stdlib/js/hull/kv/_memstore.js
/*
 * hull:kv:_memstore - native in-process byte store (JS mirror of
 * hull.kv._memstore). One physical store, two policies: hull.kv memory
 * (eviction off) and hull.cache memory (LRU eviction on). Byte-string keys and
 * values; lazy TTL expiry; byte + item accounting; LRU by access sequence.
 * Single-threaded (event-loop affinity); no locking. Stores are keyed by
 * namespace at module level, so same-namespace opens share state.
 *
 * Internal module. SPDX-License-Identifier: AGPL-3.0-or-later
 */

import util from "hull:kv:_util";

const REGISTRY = new Map(); // namespace -> Store
const ENTRY_OVERHEAD = 48;

class Store {
    constructor(policy) {
        this.data = new Map(); // key -> { v, exp(ms|null), bytes, seq }
        this.seq = 0;
        this.items = 0;
        this.bytes = 0;
        this.maxBytes = util.checkCount(policy.maxBytes, "max_bytes") || 0;
        this.maxItems = util.checkCount(policy.maxItems, "max_items") || 0;
        this.defaultTtl = policy.defaultTtl;
        this.evict = !!policy.evict;
        this.st = { hits: 0, misses: 0, evictions: 0, expirations: 0 };
        this.caps = {
            ttl: true, atomic_increment: true, compare_exchange: true,
            scan: true, persistent: false, shared: false,
            eviction: !!policy.evict, transactions: false,
        };
    }

    _drop(k, e) { this.data.delete(k); this.items -= 1; this.bytes -= e.bytes; }

    _live(k) {
        const e = this.data.get(k);
        if (!e) return null;
        if (e.exp !== null && util.nowMs() >= e.exp) {
            this._drop(k, e);
            this.st.expirations += 1;
            return null;
        }
        return e;
    }

    _makeRoom(addBytes, addingItem) {
        const over = () =>
            (this.maxItems > 0 && this.items + (addingItem ? 1 : 0) > this.maxItems) ||
            (this.maxBytes > 0 && this.bytes + addBytes > this.maxBytes);
        if (!over()) return;
        if (!this.evict)
            util.error("capacity_exceeded", "kv: in-memory store is full (eviction disabled)");
        while (over() && this.items > 0) {
            let lruK, lruSeq;
            for (const k of this.data.keys()) {
                const e = this.data.get(k);
                if (lruSeq === undefined || e.seq < lruSeq) { lruK = k; lruSeq = e.seq; }
            }
            if (lruK === undefined) break;
            this._drop(lruK, this.data.get(lruK));
            this.st.evictions += 1;
        }
        if (over())
            util.error("capacity_exceeded", "kv: value larger than the cache byte budget");
    }

    get(k) {
        const e = this._live(k);
        if (!e) { this.st.misses += 1; return null; }
        this.seq += 1; e.seq = this.seq;
        this.st.hits += 1;
        return e.v;
    }

    has(k) { return this._live(k) !== null; }

    put(k, v, ttl) {
        const exp = util.expiryMs(ttl, this.defaultTtl);
        const nb = k.length + v.length + ENTRY_OVERHEAD;
        const old = this.data.get(k);
        if (old) {
            // Bump to MRU first so make_room never evicts the key being updated.
            this.seq += 1; old.seq = this.seq;
            const delta = nb - old.bytes;
            if (delta > 0) this._makeRoom(delta, false);
            this.bytes = this.bytes - old.bytes + nb;
            old.v = v; old.exp = exp; old.bytes = nb;
            return v;
        }
        this._makeRoom(nb, true);
        this.seq += 1;
        this.data.set(k, { v, exp, bytes: nb, seq: this.seq });
        this.items += 1; this.bytes += nb;
        return v;
    }

    del(k) {
        const e = this.data.get(k);
        if (!e) return false;
        this._drop(k, e);
        return true;
    }

    incr(k, by, ttl) {
        const e = this._live(k);
        if (e) {
            // Preserve the existing expiry (matches the SQL backend and Redis
            // INCR); only the value changes.
            const newv = String(util.toInt(e.v) + by);
            const nb = k.length + newv.length + ENTRY_OVERHEAD;
            const delta = nb - e.bytes;
            this.seq += 1; e.seq = this.seq;
            if (delta > 0) this._makeRoom(delta, false);
            this.bytes = this.bytes - e.bytes + nb;
            e.v = newv; e.bytes = nb;
            return util.toInt(newv);
        }
        this.put(k, String(by), ttl); // fresh key: apply ttl
        return by;
    }


    // compare-and-swap; expected undefined/null = set only if absent.
    cas(k, expected, newVal, ttl) {
        const e = this._live(k);
        const cur = e ? e.v : null;
        const exp = (expected === undefined || expected === null) ? null : expected;
        if (cur !== exp) return false;
        this.put(k, newVal, ttl);
        return true;
    }

    clear() { this.data = new Map(); this.items = 0; this.bytes = 0; }

    scan(prefix, limit) {
        prefix = prefix || "";
        const out = [];
        for (const k of this.data.keys()) {
            const e = this.data.get(k);
            if ((e.exp === null || util.nowMs() < e.exp) && k.startsWith(prefix)) {
                out.push(k);
                if (limit && out.length >= limit) break;
            }
        }
        return out;
    }

    cleanup() {
        const now = util.nowMs();
        let removed = 0;
        for (const k of this.data.keys()) {
            const e = this.data.get(k);
            if (e.exp !== null && now >= e.exp) { this._drop(k, e); removed += 1; }
        }
        this.st.expirations += removed;
        return removed;
    }

    stats() {
        return {
            hits: this.st.hits, misses: this.st.misses,
            evictions: this.st.evictions, expirations: this.st.expirations,
            items: this.items, bytes: this.bytes,
        };
    }
}

function get(namespace, policy) {
    let s = REGISTRY.get(namespace);
    if (s) return s;
    s = new Store(policy || {});
    REGISTRY.set(namespace, s);
    return s;
}

function _forget(namespace) { REGISTRY.delete(namespace); }

export default { get, _forget };

File: stdlib/lua/hull/kv/_memstore.lua
--[[
  hull.kv._memstore - native in-process byte store.

  One physical store backs two policies: hull.kv memory (eviction OFF, KV
  semantics) and hull.cache memory (LRU eviction ON, cache semantics). Keys
  and values are arbitrary bytes (Lua strings, binary-safe as table keys).
  Lazy TTL expiry on access, byte + item accounting, LRU by access sequence.
  Single-threaded (event-loop affinity); no locking.

  Stores are keyed by namespace at MODULE level, so two opens of the same
  namespace in one process share state (in-process "shared"); policy is fixed
  by the first open of a namespace.

  Internal module. SPDX-License-Identifier: AGPL-3.0-or-later
]]

local u = require("hull.kv._util")

local M = {}
local Store = {}
Store.__index = Store

-- namespace -> Store (process-wide, so same-namespace opens share)
local REGISTRY = {}

local ENTRY_OVERHEAD = 48 -- rough per-entry bookkeeping charged to byte accounting

-- policy: { max_bytes=0, max_items=0, default_ttl=nil, evict=false }
function M.get(namespace, policy)
    local s = REGISTRY[namespace]
    if s then return s end
    policy = policy or {}
    s = setmetatable({
        data        = {},   -- key -> { v, exp(ms|nil), bytes, seq }
        seq         = 0,
        items       = 0,
        bytes       = 0,
        max_bytes   = u.check_count(policy.max_bytes, "max_bytes") or 0,
        max_items   = u.check_count(policy.max_items, "max_items") or 0,
        default_ttl = policy.default_ttl,
        evict       = policy.evict and true or false,
        st          = { hits = 0, misses = 0, evictions = 0, expirations = 0 },
        caps        = {
            ttl = true, atomic_increment = true, compare_exchange = true,
            scan = true, persistent = false, shared = false,
            eviction = policy.evict and true or false, transactions = false,
        },
    }, Store)
    REGISTRY[namespace] = s
    return s
end

local function drop(self, k, e)
    self.data[k] = nil
    self.items = self.items - 1
    self.bytes = self.bytes - e.bytes
end

-- Return the entry if live; lazily drop + return nil if expired.
local function live(self, k)
    local e = self.data[k]
    if not e then return nil end
    if e.exp and u.now_ms() >= e.exp then
        drop(self, k, e)
        self.st.expirations = self.st.expirations + 1
        return nil
    end
    return e
end

-- Evict least-recently-used live entries until the incoming entry fits, or
-- (eviction disabled) raise once a cap would be exceeded.
local function make_room(self, add_bytes, adding_item)
    local function over()
        return (self.max_items > 0 and self.items + (adding_item and 1 or 0) > self.max_items)
            or (self.max_bytes > 0 and self.bytes + add_bytes > self.max_bytes)
    end
    if not over() then return end
    if not self.evict then
        u.error("capacity_exceeded", "kv: in-memory store is full (eviction disabled)")
    end
    -- Evict lowest-seq (LRU) entries until within caps or empty.
    while over() and self.items > 0 do
        local lru_k, lru_seq
        for k, e in pairs(self.data) do
            if not lru_seq or e.seq < lru_seq then lru_k, lru_seq = k, e.seq end
        end
        if not lru_k then break end
        drop(self, lru_k, self.data[lru_k])
        self.st.evictions = self.st.evictions + 1
    end
    if over() then
        -- A single value larger than the whole budget.
        u.error("capacity_exceeded", "kv: value larger than the cache byte budget")
    end
end

function Store:get(k)
    local e = live(self, k)
    if not e then self.st.misses = self.st.misses + 1; return nil end
    self.seq = self.seq + 1; e.seq = self.seq
    self.st.hits = self.st.hits + 1
    return e.v
end

function Store:has(k) return live(self, k) ~= nil end

function Store:put(k, v, ttl)
    local exp = u.expiry_ms(ttl, self.default_ttl)
    local nb = #k + #v + ENTRY_OVERHEAD
    local old = self.data[k]
    if old then
        -- Overwrite: bump to MRU first so make_room never evicts the key we are
        -- updating, then account only the byte delta.
        self.seq = self.seq + 1
        old.seq = self.seq
        local delta = nb - old.bytes
        if delta > 0 then make_room(self, delta, false) end
        self.bytes = self.bytes - old.bytes + nb
        old.v, old.exp, old.bytes = v, exp, nb
        return v
    end
    make_room(self, nb, true)
    self.seq = self.seq + 1
    self.data[k] = { v = v, exp = exp, bytes = nb, seq = self.seq }
    self.items = self.items + 1
    self.bytes = self.bytes + nb
    return v
end

function Store:del(k)
    local e = self.data[k]
    if not e then return false end
    drop(self, k, e)
    return true
end

function Store:incr(k, by, ttl)
    local e = live(self, k)
    if e then
        -- Preserve the existing expiry (matches the SQL backend and Redis
        -- INCR); only the value changes.
        local newv = tostring(u.to_int(e.v) + by)
        local nb = #k + #newv + ENTRY_OVERHEAD
        local delta = nb - e.bytes
        self.seq = self.seq + 1
        e.seq = self.seq
        if delta > 0 then make_room(self, delta, false) end
        self.bytes = self.bytes - e.bytes + nb
        e.v, e.bytes = newv, nb
        return u.to_int(newv)
    end
    self:put(k, tostring(by), ttl)   -- fresh key: apply ttl
    return by
end

-- compare-and-swap: expected may be nil (set only if absent). Returns bool.
function Store:cas(k, expected, new, ttl)
    local e = live(self, k)
    local cur = e and e.v or nil
    if cur ~= expected then return false end
    self:put(k, new, ttl)
    return true
end

function Store:clear()
    self.data, self.items, self.bytes = {}, 0, 0
end

function Store:scan(prefix, limit)
    prefix = prefix or ""
    local out = {}
    for k, e in pairs(self.data) do
        if (not e.exp or u.now_ms() < e.exp) and k:sub(1, #prefix) == prefix then
            out[#out + 1] = k
            if limit and #out >= limit then break end
        end
    end
    return out
end

-- Sweep expired entries eagerly (optional hygiene; access already lazy-expires).
function Store:cleanup()
    local now, removed = u.now_ms(), 0
    for k, e in pairs(self.data) do
        if e.exp and now >= e.exp then drop(self, k, e); removed = removed + 1 end
    end
    self.st.expirations = self.st.expirations + removed
    return removed
end

function Store:stats()
    return {
        hits = self.st.hits, misses = self.st.misses,
        evictions = self.st.evictions, expirations = self.st.expirations,
        items = self.items, bytes = self.bytes,
    }
end

-- Test/lifecycle helper: forget a namespace's store entirely.
function M._forget(namespace) REGISTRY[namespace] = nil end

return M
